enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,169 @@
+
+#### Backend - Integrated OAuth Server Token Authentication
+
+1. cluster-authentication-operator observes `authentication.config/cluster` `type` field
+2. if set to "IntegratedOAuth", it creates an `openshift-config/webhook-authentication-integrated-oauth`
+  secret and adds it to the `webhookTokenAuthenticators` field of the `authentication.config/cluster`
+  resource
+  - the above secret contains details that point kube-apiserver to the oauth-apiserver for token reviews
+3. during an authentication flow, kube-apiserver sends `tokenreviews.authentication.k8s.io` object to the
+  oauth-apiserver for review
+4. oauth-apiserver sends a response accordingly to Kubernetes requirements, on success it adds the
+  `system:authenticated:oauth` group to the returned details about the groups of the authenticated user
+  - `system:authenticated:oauth` possesses privileges for users to handle projects via the self-provisioner
+    cluster role
+  - similarly, if any integrator would like their users to have project privileges, they need to add this
+    specific group in their token review responses.
+
+### Common Gotchas When Providing Your Own Authentication Webhook
+
+1. `oauth-proxy` relies on the _/.well-known/oauth-authorization-server_ endpoint
+   and if the integrated oauth-server is turned off and the `spec.oauthMetadata` field
+   in `authentications.config/cluster` is not set   properly, things like grafana
+   and prometheus won't be accessible
+2. ServiceAccounts may not be usable as OAuth clients anymore
+
+It is a sole responsibility of the replacing part to make sure all of these work.
+
+## Design Details
+
+### Test Plan
+
+Since we will be making the cluster-authentication-operator use the `webhookTokenAuthenticatros`
+(in combination with `oauthMetadata`, which already works) to configure authentication using
+the integrated OAuth server, we will just re-use all the tests that we have already.
+
+### Upgrade / Downgrade Strategy
+
+To keep authentication working even during upgrades, we'll keep our kube-apiserver
+authenticator patch code in place for 4.n version, but will disable it if any webhook
+authenticators are set. The authenticator patching code should then be removed in 4.(n+1)
+version.
+
+The cluster-authentication-operator in 4.n will wait for the kube-apiserver-operator to
+announce its 4.n version, and once that happens, and only after the oauth-apiserver deployment
+is healthy, it will create the `openshift-config/webhook-authentication-integrated-oauth`
+secret and configure the `webhookTokenAuthenticators` field of the `authentications.config/cluster`.
+
+### Version Skew Strategy
+
+See the upgrade/downgrade strategy.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
```

Comment:
I would also love to see a an example, how to hook a custom IdP such as Keycloak. From the design, I have an impression that we'll need to:

- Create a Secret in `openshift-config` containing a `kubeconfig` key. The value will look like a typical `Config`:

```yaml
apiVersion: v1
kind: Config
users:
- name: <Keycloak Service Account>
  user:
    auth-provider:
      config:
         idp-issuer-url: <Keycloak URL> 
         idp-client-id: <Keycloak Client ID>
         idp-client-secret: <Keycloak Client Secret>
         idp-client-secret: <Keycloak Client Secret>
         idp-certificate-authority: <This one is very tricky - it should be a path, but we probably want to use a secret here>
clusters:
- cluster:
    certificate-authority-data: <CA Bundle used for connecting to the API Server>
    # Other fields used for connecting to an OpenShift cluster
  name: openshift
contexts:
- context:
    cluster: openshift
    user: <Keycloak Service Account>
  name: openshift
current-context: openshift
```

* We add the secret reference to `webhookTokenAuthenticators` field of the `authentication.config/cluster` resource.

I'm very interested in the structure of the Secret used in `openshift-config` namespace. This is how I understood the design but I'm almost sure, I missed something...

Reply:
@stlaz Could you please tell me how the documentation for plugging custom IdP in will look like?

Reply:
sure, briefly put:
1. create a config file in the form of:
```
apiVersion: v1
kind: Config
clusters:
  - name: my-webhook-server
    cluster:
      certificate-authority-data: <base-64 encoded PEM CA cert>
      server: https://authn.my-webhook.com/authenticate # URL of remote service to query. Must use 'https'.

# users refers to the API server's webhook configuration.
users:
  - name: openshift-kube-apiserver
    user:
      client-certificate-data: <base64 encoded client-auth cert> # cert for the webhook plugin to use
      client-key-data: <base64-encoded key for the cert>          # key matching the cert

# kubeconfig files require a context. Provide one for the API server.
current-context: webhook
contexts:
- context:
    cluster: my-webhook-server
    user: openshift-kube-apiserver
  name: webhook
```
2. `oc create secret generic -n openshift-config auth-webhook-kubeconfig --from-file=kubeconfig=<path to file from 1.>`
3. edit the `type` and `webhookTokenAuthenticator` field of the `authentication.config.openshift.io/cluster` resource so that `type` is set to `None` and `webhookTokenAuthenticator` is set to `auth-webhook-kubeconfig` (the name of the secret from step 2)

Reply:
4. create a file for OIDC discovery as described in https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
5. `oc create cm -n openshift-config --from-file=oauthMetadata=<path to file from 4.>`
6. configure the `oauthMetadata` field of the `authentication.config.openshift.io/cluster` resource to reference the configMap created in step 5.

Reply:
3. should actually be in 6., I originally forgot about the oauth-metadata part

Reply:
Thanks a lot @stlaz !

Regarding point number 4 - this is provided by Keycloak with an URL. Here's an example: https://gist.github.com/slaskawi/ce519d994a166bbb7877a94069ea49f6

Perhaps we could design it in such a way, so that we could point OpenShift directly to this URL (without creating a file)?

Reply:
I think delegating the endpoint could make sense, although it's not how this works today, unfortunately. We would need a different enhancement for that. I'll give it a thought.